example/hyperparam/search_gpyopt.py

```diff
@@ -0,0 +1,204 @@
+                valid_loss = min(null_valid_loss,
+                                 store.get_metric(p.run_id, "val_{}".format(metric)).value)
+                test_loss = min(null_test_loss,
+                                store.get_metric(p.run_id, "test_{}".format(metric)).value)
+
+            else:
+                # run failed => return null loss
+                tracking_service.set_terminated(p.run_id, "FAILED")
+                train_loss = null_train_loss
+                valid_loss = null_valid_loss
+                test_loss = null_test_loss
+
+            mlflow.log_metric("train_{}".format(metric), valid_loss)
+            mlflow.log_metric("val_{}".format(metric), valid_loss)
+            mlflow.log_metric("test_{}".format(metric), test_loss)
+            with open(results_path, "a") as f:
+                f.write("{runId} {train} {val} {test}\n".format(runId=p.run_id,
+                                                                train=train_loss,
+                                                                val=valid_loss,
+                                                                test=test_loss))
+            if return_all:
+                return train_loss, valid_loss, test_loss
+            else:
+                return valid_loss
+
+        return eval
+
+    with mlflow.start_run() as run:
+        experiment_id = run.info.experiment_id if training_experiment_id == -1 \
+            else training_experiment_id
+        # Evaluate null model first.
+        # We use null model (predict everything to the mean) as a reasonable upper bound on loss.
+        # We need an upper bound to handle the failed runs (e.g. return NaNs) because GPyOpt can not
+        # handle Infs.
+        # Allways including a null model in our results is also a good ML practice.
+        train_null_loss, valid_null_loss, test_null_loss = new_eval(0,
+                                                                    experiment_id,
+                                                                    math.inf,
+                                                                    math.inf,
+                                                                    math.inf,
+                                                                    True)(params=[[0, 0]])
+        myProblem = GPyOpt.methods.BayesianOptimization(new_eval(epochs,
+                                                                 experiment_id,
+                                                                 train_null_loss,
+                                                                 valid_null_loss,
+                                                                 test_null_loss),
+                                                        bounds,
+                                                        batch_size=batch_size,
+                                                        num_cores=max_p,
```

Comment:
BTW this was with batch_size=4 and max_p=4. There were two parallel runs at the start but then everything is serial.

Reply:
That is because the default evaluator for GPyOpt is sequential. This means that the first n random samples are evaluated in parallel but after that the models are evaluated one by one. I will change this to evaluator which works in parallel.